### ISSUE TITLE
Give a convenient way to retrieve shadow variables created by ExponentialMovingAverage.

### ISSUE THREAD

**Issue by panmari** [tensorflow/tensorflow#823]
At first I was under the impression, the variables returned by `tf.moving_average_variables()` would be shadow variables created, but that does not seem to be the case (they didn't save/restore correctly for me). 

Now that I'm reading the code, that seems to be the intention, though? Shouldn't [this line](https://github.com/tensorflow/tensorflow/blame/f2bd0fc399606d14b55f3f7d732d013f32b33dd5/tensorflow/python/training/moving_averages.py#L227) read 

```
ops.add_to_collection(ops.GraphKeys.MOVING_AVERAGE_VARIABLES, avg)
```

? 

If not, it would be great to save the averages under some other key and make them easily retrievable for the saver in a similar way as `tf.moving_average_variables()`, maybe as  `tf.moving_average_shadow_variables()`.

**Comment by josh11b** [comment]
That could very well be a bug.  Sherry, could you look at this?

**Comment by panmari** [comment]
Ping @sherrym, could you please give your opinion?

**Comment by sherrym** [comment]
HI there,

Sorry about the delay.

What did you mean by "they didn't save/restore correctly for me"?

We have introduced variables_to_restore() (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/moving_averages.py#L327) to return a map from the moving_average_names to the variables mappings, such that you can restore the moving_average values into the corresponding variables. Would that do what you want? If not, would you please elaborate what you would like to restore?

Thanks,
Sherry

**Comment by aplatanado** [comment]
Hi,

I guess I understand what @panmari says because I have found the same problem.

There are no way to obtain the list of shadow variables, for example, for just save them.

After read the documentation, tf.moving_average_variables() seems to do the work. But it isn't because tf.moving_average_variables() returns the variables which moving average is applied not its shadow variables.

Thanks

**Comment by girving** [comment]
@martinwicke, @aselle: I think we may need backup czars, since Sherry is out at the moment.

**Comment by suharshs** [comment]
@sherrym

**Comment by bsautermeister** [comment]
Hi, is there any progress already?

I'm currently facing the same issue. When I save and restore my trained model as explained [here](https://www.tensorflow.org/versions/r0.10/how_tos/variables/index.html) and I would like to continue my training, I was always wondering why the restored model always performs slightly worse that before saving it. I have the impression that restore() is not actually restoring the shadow-variables.

In case we have to save/restore them manually, that would be OK for me, but could you tell an example. Because I cannot use mae.variables_to_restore(), because I cannot access the ExponentialMovingAverage instance, which has been written in a different module.

Beside ExponentialMovingAverage, I have the same problem for the shadow-variables within the Optimizers (such as Adam).

**Comment by martinwicke** [comment]
So that would mean this bug is done? Or is there an issue still hidden here? If so, please comment to reopen, or file a new, specific issue.

**Comment by aplatanado** [comment]
The issue is that there is no function to retrieve only the shadow variables. Something like tf.moving_average_shadow_variables(). Because tf.moving_average_variables() returns the variables which moving average is applied.

**Comment by saminkhan** [comment]
Is there any best practice yet to retrieve shadow variable mappings for inference time?

Here's a custom function I'm using to get shadow variable mappings instead of regular ones (returns regular variable mapping if shadow variable doesn't exist)....

> 
    def get_shadow_variables(decay_r):
        names_set, shadow_vars = set(), {}
    
        # assumption: shadow variable names appear after
        # regular variable names in alphabetical list
        sorted_items = sorted(
            tf.train.ExponentialMovingAverage(
                decay_r).variables_to_restore().items(), reverse=True)
    
        for key, val in sorted_items:
            if val.name not in names_set:
                names_set.add(val.name)
                shadow_vars[key] = val
    
        return shadow_vars

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by bhack** [comment]
Is this still valid? `tf.moving_average_variables` Is a TF v1 only API now.

**Comment by panmari** [comment]
I guess this is by now obsolete.
